### PR TITLE
docs: add documentation for assert API (#4242)

### LINKED ISSUE #4242: Document "assert"
### Clear and concise description of the problem

Vitest exports `assert` method which is not documented anywhere. It is just reexport of https://www.chaijs.com/api/assert/

### Suggested solution

Make our own documentation of this method.

### Alternative

_No response_

### Additional context

_No response_

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't already an issue that request the same feature to avoid creating a duplicate.

### PR DESCRIPTION
**Pull request by willieho**
### Description

Closes #4242 
Initiate the doc for `assert`

screencapture-localhost-3333-api-assert-html-2024-02-06-22_51_31

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. If the feature is substantial or introduces breaking changes without a discussion, PR might be closed.
- [ ] Ideally, include a test that fails without this PR but passes with it.
- [x] Please, don't make changes to `pnpm-lock.yaml` unless you introduce a new test example.

### Tests
- [ ] Run the tests with `pnpm test:ci`.

### Documentation
- [ ] If you introduce new functionality, document it. You can run documentation with `pnpm run docs` command.

### Changesets
- [ ] Changes in changelog are generated from PR name. Please, make sure that it explains your changes in an understandable manner. Please, prefix changeset messages with `feat:`, `fix:`, `perf:`, `docs:`, or `chore:`.

### CODE REVIEW

**Comment by developer-bandi** [comment]
@willieho 
i finish this job but i don't know how to add my commit in this branch 
One way I figured out is to create a new pr containing your task and close this pr.
What should I do?

**Comment by willieho** [comment]
@developer-bandi Sorry for late reply, you can refer to [this instruction](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/committing-changes-to-a-pull-request-branch-created-from-a-fork).

**Comment by developer-bandi** [comment]
> @developer-bandi Sorry for late reply, you can refer to [this instruction](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/committing-changes-to-a-pull-request-branch-created-from-a-fork), and it basically is "fork the fork".

thank you

However, I cloned the repository you forked using the link above, added a commit to the doc/add-doc-for-assert branch, and pushed, but the push does not work.

Is there a way? The error message is attached below.

스크린샷 2024-02-16 오후 12 38 11

**Comment by willieho** [comment]
@developer-bandi  I believe the correct way is to fork my forked repo, add commit, and open a new PR (but to be honest, I didn't try it myself). In that case, I think my commit will be contained and you will have full control of it.

It should work in the similar way If you want to push to my forked repo (I just added you as a collaborator), this way we can keep using this PR but it will need us to work together if the team has any feedback or need any adjustments.

If above options are not ok to you (or they don't work), I don't mind if you just create your own PR like you mentioned at the beginning. 
Again, sorry for the inconvenience.

**Comment by developer-bandi** [comment]
@willieho Thanks to adding me as a collaborator and detailed explanation, I was able to add the commit to mr by making a push request.

After reviewing, if there are no problems, please change the draft to mergeable!

File: docs/api/assert.md [hunk 1]
```diff
@@ -0,0 +1,1960 @@
...
+test('assert.isObject', () => {
+  assert.isObject(someThing, 'someThing is object')
+})
+```
+
+## isNotObject
+
+- **Type:** `<T>(value: T, message?: string) => void`
+
+Asserts that `value` is an object of type Object (as revealed by Object.prototype.toString). The assertion does not match subclassed objects.
```
Comment (sheremet-va) [inline docs/api/assert.md:443]:
```suggestion
Asserts that `value` is not an object of type Object (as revealed by Object.prototype.toString). The assertion does not match subclassed objects.
```

File: docs/api/assert.md [hunk 2]
```diff
@@ -0,0 +1,1960 @@
...
+test('assert.isNotNumber', () => {
+  assert.isNotNumber(colors, 'colors is not number but strings')
+})
+```
+
+## isFinite
+
+- **Type:** `<T>(value: T, message?: string) => void`
+
+Asserts that `value` is a finite number(not NaN, Infinity).
```
Comment (sheremet-va) [inline docs/api/assert.md:555]:
```suggestion
Asserts that `value` is a finite number (not NaN, Infinity).
```

File: docs/api/assert.md [hunk 3]
```diff
@@ -0,0 +1,1960 @@
...
+  assert.instanceOf(foo, Person, 'foo is an instance of Person')
+  assert.instanceOf(coffee, Tea, 'coffee is an instance of Tea')
+})
+```
+
+## notInstanceOf
+
+- **Type:** `<T>(value: T, constructor: Function, message?: string) => void`
+
+Asserts that `value` is an instance of `constructor`.
```
Comment (sheremet-va) [inline docs/api/assert.md:661]:
```suggestion
Asserts that `value` is not an instance of `constructor`.
```

File: docs/api/assert.md [hunk 4]
```diff
@@ -0,0 +1,1960 @@
...
+```
+
+## notDeepInclude
+
+- **Type:**
+  - `(haystack: string, needle: string, message?: string) => void`
+  - `<T>(haystack: readonly T[] | ReadonlySet<T> | ReadonlyMap<any, T>, needle: T, message?: string) => void`
+  - `<T>(haystack: T, needle: T extends WeakSet<any> ? never : Partial<T>, message?: string) => void`
+
+Asserts that `haystack` includes `needle`. Can be used to assert the absence of a value in an array or a subset of properties in an object. Deep equality is used.
```
Comment (sheremet-va) [inline docs/api/assert.md:749]:
```suggestion
Asserts that `haystack` does not include `needle`. Can be used to assert the absence of a value in an array or a subset of properties in an object. Deep equality is used.
```

File: docs/api/assert.md [hunk 5]
```diff
@@ -0,0 +1,1960 @@
...
+  assert.nestedInclude({ '.a': { b: 'x' } }, { '\\.a.[b]': 'x' })
+  assert.nestedInclude({ a: { '[b]': 'x' } }, { 'a.\\[b\\]': 'x' })
+})
+```
+
+## notNestedInclude
+
+- **Type:** `(haystack: any, needle: any, message?: string) => void`
+
+Asserts that `haystack` does not includes `needle`. Can be used to assert the inclusion of a subset of properties in an object. Enables the use of dot- and bracket-notation for referencing nested properties. ‘[]’ and ‘.’ in property names can be escaped using double backslashes.
```
Comment (sheremet-va) [inline docs/api/assert.md:782]:
```suggestion
Asserts that `haystack` does not include `needle`. Can be used to assert the inclusion of a subset of properties in an object. Enables the use of dot- and bracket-notation for referencing nested properties. ‘[]’ and ‘.’ in property names can be escaped using double backslashes.
```

File: docs/api/assert.md [hunk 6]
```diff
@@ -0,0 +1,1960 @@
...
+  assert.deepNestedInclude({ a: { b: [{ x: 1 }] } }, { 'a.b[0]': { x: 1 } })
+  assert.deepNestedInclude({ '.a': { '[b]': { x: 1 } } }, { '\\.a.\\[b\\]': { x: 1 } })
+})
+```
+
+## notDeepNestedInclude
+
+- **Type:** `(haystack: any, needle: any, message?: string) => void`
+
+Asserts that `haystack` not includes `needle`. Can be used to assert the absence of a subset of properties in an object while checking for deep equality. Enables the use of dot- and bracket-notation for referencing nested properties. ‘[]’ and ‘.’ in property names can be escaped using double backslashes.
```
Comment (sheremet-va) [inline docs/api/assert.md:812]:
```suggestion
Asserts that `haystack` does not include `needle`. Can be used to assert the absence of a subset of properties in an object while checking for deep equality. Enables the use of dot- and bracket-notation for referencing nested properties. ‘[]’ and ‘.’ in property names can be escaped using double backslashes.
```

File: docs/api/assert.md [hunk 7]
```diff
@@ -0,0 +1,1960 @@
...
+test('assert.ownInclude', () => {
+  assert.ownInclude({ a: 1 }, { a: 1 })
+})
+```
+
+## notOwnInclude
+
+- **Type:** `(haystack: any, needle: any, message?: string) => void`
+
+Asserts that `haystack` includes `needle`. Can be used to assert the absence of a subset of properties in an object while ignoring inherited properties.
```
Comment (sheremet-va) [inline docs/api/assert.md:841]:
```suggestion
Asserts that `haystack` doesn't include `needle`. Can be used to assert the absence of a subset of properties in an object while ignoring inherited properties.
```

File: docs/api/assert.md [hunk 8]
```diff
@@ -0,0 +1,1960 @@
...
+test('assert.deepOwnInclude', () => {
+  assert.deepOwnInclude({ a: { b: 2 } }, { a: { b: 2 } })
+})
+```
+
+## notDeepOwnInclude
+
+- **Type:** `(haystack: any, needle: any, message?: string) => void`
+
+Asserts that `haystack` not includes `needle`. Can be used to assert the absence of a subset of properties in an object while ignoring inherited properties and checking for deep equality.
```
Comment (sheremet-va) [inline docs/api/assert.md:876]:
```suggestion
Asserts that `haystack` does not include `needle`. Can be used to assert the absence of a subset of properties in an object while ignoring inherited properties and checking for deep equality.
```

File: docs/api/assert.md [hunk 9]
```diff
@@ -0,0 +1,1960 @@
...
+test('assert.match', () => {
+  assert.match('foobar', /^foo/, 'regexp matches')
+})
+```
+
+## notMatch
+
+- **Type:** `(value: string, regexp: RegExp, message?: string) => void`
+
+Asserts that `value` does not matches the regular expression `regexp`.
```
Comment (sheremet-va) [inline docs/api/assert.md:904]:
```suggestion
Asserts that `value` does not matche the regular expression `regexp`.
```

File: docs/api/assert.md [hunk 10]
```diff
@@ -0,0 +1,1960 @@
...
+
+- **Type:** `<T>(object: T, property: string, message?: string) => void`
+
+Asserts that `object` has a direct or inherited property named by `property`
+
+```ts
+import { assert, test } from 'vitest'
+
+test('assert.property', () => {
+  assert.property({ tea: { green: 'matcha' } })
```
Comment (sheremet-va) [inline docs/api/assert.md:924]: these examples are missing the second argument

File: docs/api/assert.md [hunk 11]
```diff
@@ -0,0 +1,1960 @@
...
+
+- **Type:** `<T>(object: T, property: string, message?: string) => void`
+
+Asserts that `object` does not have a direct or inherited property named by `property`
+
+```ts
+import { assert, test } from 'vitest'
+
+test('assert.notProperty', () => {
+  assert.notProperty({ tea: { green: 'matcha' } })
```
Comment (sheremet-va) [inline docs/api/assert.md:939]: and this one

File: docs/api/assert.md [hunk 12]
```diff
@@ -0,0 +1,1960 @@
...
+
+- **Type:** `<T, V>(object: T, property: string, value: V, message?: string) => void`
+
+Asserts that `object` has a direct or inherited property named by `property` with a value given by `value`. Uses a strict equality check (===).
+
+```ts
+import { assert, test } from 'vitest'
+
+test('assert.notPropertyVal', () => {
+  assert.propertyVal({ tea: 'is good' })
```
Comment (sheremet-va) [inline docs/api/assert.md:953]: and this one

File: docs/api/assert.md [hunk 13]
```diff
@@ -0,0 +1,1960 @@
...
+
+- **Type:** `<T, V>(object: T, property: string, value: V, message?: string) => void`
+
+Asserts that `object` does not have a direct or inherited property named by `property` with a value given by `value`. Uses a strict equality check (===).
+
+```ts
+import { assert, test } from 'vitest'
+
+test('assert.notPropertyVal', () => {
+  assert.notPropertyVal({ tea: 'is good' })
```
Comment (sheremet-va) [inline docs/api/assert.md:967]: and these

File: docs/api/assert.md [hunk 14]
```diff
@@ -0,0 +1,1960 @@
...
+import { assert, test } from 'vitest'
+
+test('assert.deepPropertyVal', () => {
+  assert.notNestedProperty({ tea: { green: 'matcha' } }, 'tea.oolong')
+})
+```
+
+## nestedPropertyVal
+
+- **Type:** `<T>(object: T, property: string, value: any, message?: string) => void`
```
Comment (sheremet-va) [inline docs/api/assert.md:1032]: is value required? example doesn't show it

**Review by sheremet-va — CHANGES_REQUESTED** [review]
The PR is too big to review it whole, but I have a few comments already 😄

File: docs/api/assert.md (hunk 14 above)
Comment (developer-bandi) [reply docs/api/assert.md:1032]: Yes, the type definition is correct, but the value was omitted from the example 

so i add value in example again

**Comment by developer-bandi** [comment]
> The PR is too big to review it whole, but I have a few comments already 😄

Thanks for the review!

File: docs/api/assert.md [hunk 15]
```diff
@@ -0,0 +1,1960 @@
...
+  function fn() { obj.val = 5 };
+  assert.doesNotDecreaseBy(fn, obj, 'val', 1)
+})
+```
+
+## decreasesButNotBy
+
+- **Type:** `<T>(modifier: Function, object: T, property: string, change: number, message?: string) => void`
+
+Asserts that a `modifier` does not decreases a numeric `object`'s `property` or a `modifier` return value by an `change`.
```
Comment (dammy001) [inline docs/api/assert.md:1798]:
```suggestion
Asserts that a `modifier` does not decrease a numeric `object`'s `property` or a `modifier` return value by a `change`.
```

File: docs/api/assert.md [hunk 16]
```diff
@@ -0,0 +1,1960 @@
...
+  assert.isNotFrozen({})
+})
+```
+
+## isEmpty
+
+- **Type:** `<T>(target: T, message?: string) => void`
+- **Alias:** `empty`
+
+Asserts that the `target` does not contain any values. For arrays and strings, it checks the length property. For Map and Set instances, it checks the size property. For non-function objects, it gets the count of own enumerable string keys.
```
Comment (dammy001) [inline docs/api/assert.md:1931]:
```suggestion
Asserts that the `target` does not contain any values. For arrays and strings, it checks the length property. For Map and Set instances, it checks the size property. For non-function objects, it gets the count of its own enumerable string keys.
```

File: docs/api/assert.md [hunk 17]
```diff
@@ -0,0 +1,1960 @@
...
+  assert.isEmpty({})
+})
+```
+
+## isNotEmpty
+
+- **Type:** `<T>(object: T, message?: string) => void`
+- **Alias:** `notEmpty`
+
+Asserts that the `target` contains values. For arrays and strings, it checks the length property. For Map and Set instances, it checks the size property. For non-function objects, it gets the count of own enumerable string keys.
```
Comment (dammy001) [inline docs/api/assert.md:1949]:
```suggestion
Asserts that the `target` contains values. For arrays and strings, it checks the length property. For Map and Set instances, it checks the size property. For non-function objects, it gets the count of its own enumerable string keys.
```

**Review by sheremet-va — APPROVED** [review]

**Pull request merged**